From fbf9f4a9b6184e3762da428a2a32da179e01c542 Mon Sep 17 00:00:00 2001
From: Yu Kuai <yukuai3@huawei.com>
Date: Wed, 26 Jan 2022 09:21:32 +0800
Subject: [PATCH 2604/4585] [Backport] blk-mq: fix missing
 blk_account_io_done() in error path

mainline inclusion
from mainline-v5.15-rc1
commit 89f871af1b26d98d983cba7ed0e86effa45ba5f8
category: bugfix
bugzilla: 187014

Reference: https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?id=89f871af1b26d98d983cba7ed0e86effa45ba5f8

--------------------------------

If blk_mq_request_issue_directly() failed from
blk_insert_cloned_request(), the request will be accounted start.
Currently, blk_insert_cloned_request() is only called by dm, and such
request won't be accounted done by dm.

In normal path, io will be accounted start from blk_mq_bio_to_request(),
when the request is allocated, and such io will be accounted done from
__blk_mq_end_request_acct() whether it succeeded or failed. Thus add
blk_account_io_done() to fix the problem.

Signed-off-by: Yu Kuai <yukuai3@huawei.com>
Reviewed-by: Christoph Hellwig <hch@lst.de>
Link: https://lore.kernel.org/r/20220126012132.3111551-1-yukuai3@huawei.com
Signed-off-by: Jens Axboe <axboe@kernel.dk>

Conflict:
	block/blk-core.c
Signed-off-by: Zhang Wensheng <zhangwensheng5@huawei.com>
---
 block/blk-core.c | 5 ++++-
 1 file changed, 4 insertions(+), 1 deletion(-)

diff --git a/block/blk-core.c b/block/blk-core.c
index 109fb2750453..a94596ec6c2e 100644
--- a/block/blk-core.c
+++ b/block/blk-core.c
@@ -1208,7 +1208,10 @@ blk_status_t blk_insert_cloned_request(struct request_queue *q, struct request *
 	 * bypass a potential scheduler on the bottom device for
 	 * insert.
 	 */
-	return blk_mq_request_issue_directly(rq, true);
+	ret = blk_mq_request_issue_directly(rq, true);
+	if (ret)
+		blk_account_io_done(rq, ktime_get_ns());
+	return ret;
 }
 EXPORT_SYMBOL_GPL(blk_insert_cloned_request);
 
-- 
2.17.1

